### PR TITLE
storage_benchmark: Add a storage benchmark class

### PR DESCRIPTION
**Pull request by YongxueHong**
1. Add iozone class which providing interfaces to
test block devices.
2. Add fio class which providing interfaces to
test block devices.

ID: 1579750, 1668969 
Signed-off-by: Yongxue Hong <yhong@redhat.com>

### CODE REVIEW

**Comment by YongxueHong** [comment]
hi @XueqiangWei  @peixiu ,  help review, thanks.

**Comment by YongxueHong** [comment]
@kimi1978wy pls help me review, thanks.

File: provider/storage_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    """
+    Create a soft symbol link inside guest.
+
+    :param test: kvm test object.
+    :param session: session connected guest.
+    :param src: source file.
+    :param dst: destination file.
+    :param timeout: timeout for creating link.
+    """
+    symbol_link_cmd = 'ln -s %s %s' % (src, dst)
```
Comment (lolyu) [inline provider/storage_benchmark.py]: why not using `-f` to force create if destination exists.

Comment (YongxueHong) [reply provider/storage_benchmark.py]: You are right, I forget the '-f' option, thank your comment, I will update it.

Comment (YongxueHong) [reply provider/storage_benchmark.py]: Updated it:

>     symbol_link_cmd = 'ln -s -f %s %s' % (src, dst)
>     status, output = session.cmd_status_output(symbol_link_cmd,
>                                                timeout=timeout)
>     if status:
>         test.error('Failed to create symbol link. %s' % output)

**Comment by YongxueHong** [comment]
Hi,  Peixiu
From your log:
```
	Command line used: /Iozone/iozone -azR -r 64k -n 1G -g 1G -M -b iozone.xls -f E:\testfile -a
	Output is in Kbytes/sec
	Time Resolution = 0.000004 seconds.
	Processor cache size set to 1024 Kbytes.
	Processor cache line size set to 32 bytes.
	File stride size set to 17 * record size.
                                                            random  random    bkwd   record   stride                                   
              KB  reclen   write rewrite    read    reread    read   write    read  rewrite     read   fwrite frewrite   fread  freread
         1048576      64cygwin warning:
  MS-DOS style path detected: E:\testfile
  Preferred POSIX equivalent is: /cygdrive/e/testfile
  CYGWIN environment variable option "nodosfilewarning" turns off this warning.
  Consult the user's guide for more details about POSIX paths:
    http://cygwin.com/cygwin-ug-net/using.html#using-pathnames
 1368940 1978651  2934906  2943818 2996976 2046661 4659832  9904750  2110838  2182704  2713731 2935996  2938485

```
Found the output data are not continuous，like this: 
```
	File stride size set to 17 * record size.
                                                            random  random    bkwd   record   stride                                   
              KB  reclen   write rewrite    read    reread    read   write    read  rewrite     read   fwrite frewrite   fread  freread
         1048576      64 2554864 3797667  4868798  4880669 3625555 2611063 3182589  8928966  3799313  3419593  3798474 3959175  3974990

```
So for parse_file in avocado-vt/virttest/postprocess_iozone.py will returen a None.
Fowlloing is this function:
```
    def parse_file(self, fileobj):
        """
        Parse an IOzone results file.

        :param file: File object that will be parsed.
        :return: Matrix containing IOzone results extracted from the file.
        """
        lines = []
        for line in fileobj.readlines():
            fields = line.split()
            if len(fields) != 15:
                continue
            try:
                lines.append([int(i) for i in fields])
            except ValueError:
                continue
        return lines
```

**Comment by YongxueHong** [comment]
Hi, Peixiu

For the above error, have fixed, please test it again with latest code, thanks.

File: provider/storage_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        analysis_path = os.path.join(self.test.resultsdir, analysis_name)
+        if post and record:
+            if not re.search(r'\s+-i\s+\S+', cmd_options):
+                self.post_result(results_path, analysis_path)
+            else:
+                logging.warn('Maybe not be analysed correctly '
+                             'since iozone commands includes \'-i\' '
+                             'option. So Cancel it')
+        if status:
+            self.test.error('Failed to execute %s. %s' % (cmd, output))
```
Comment (peixiu) [inline provider/storage_benchmark.py]: same with before, suggest add '\n' to print output message from new line. and same advice for others need print output.
like self.test.error('Failed to execute %s.\n %s' % (cmd, output))

File: provider/storage_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    :param session: session connected guest.
+    :param src: source file.
+    :param dst: destination file.
+    :param timeout: timeout for creating link.
+    """
+    symbol_link_cmd = 'ln -s -f %s %s' % (src, dst)
+    status, output = session.cmd_status_output(symbol_link_cmd,
+                                               timeout=timeout)
+    if status:
+        test.error('Failed to create symbol link. %s' % output)
```
Comment (peixiu) [inline provider/storage_benchmark.py]: suggest add '\n' in error massage, start a new line for cmd output.

**Comment by YongxueHong** [comment]
Hi, Peixiu
For the above 3 comments, I have updated it, please review it again if you are free.
Thanks for your comments.

**Comment by peixiu** [comment]
It's good to me~

(01/16) Host_RHEL.*.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (184.27 s)
 (07/12) Host_RHEL.m*.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_system_disk.q35: PASS (599.21 s)
 (08/12) Host_RHEL.m*.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (544.88 s)

Acked-by: Peixiu phou@redhat.com

File: provider/storage_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+    :param filename: The name of log file.
+    :return Results path.
+    """
+    results_path = os.path.join(test.resultsdir, filename)
+    logging.debug("Write results to %s" % results_path)
+
+    try:
+        with open(results_path, 'a') as result:
+            result.write(output)
+            return results_path
```
Comment (hereischen) [inline provider/storage_benchmark.py]: indent `return results_path` with `with open(results_path, 'a') as result:`?

File: provider/storage_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+            else:
+                logging.warn('Maybe not be analysed correctly '
+                             'since iozone commands includes \'-i\' '
+                             'option. So Cancel it')
+        return output
+
+
+class IozoneLinux(Iozone):
+    def __init__(self, test, params, vm, session):
+        self.session = vm_session_check(vm, session)
```
Comment (hereischen) [inline provider/storage_benchmark.py]: I think the session's state is better maintained outside `__init__`, that is too say when people trying to use these class a living session should be provide.

File: provider/storage_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+    Wait all the specify processes are done.
+
+    :param test: Kvm test object.
+    :param session: Session connected guest.
+    :param os_type: The os type of guest.
+    :param pid_name: str, pid name.
+    :param interval: float, Check the given processed by interval.
+    :param timeout: float, Timeout for waiting.
+    """
+    end_time = float(timeout) + time.time()
```
Comment (hereischen) [inline provider/storage_benchmark.py]: don't need `float` here as `time.time()` is float

File: provider/storage_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+        :rtype: str
+        """
+        random_string = utils_misc.generate_random_string(8)
+        stamp = time.strftime('%Y-%m-%d-%H-%M-%S') + random_string
+
+        status, output = self.session.cmd_status_output(cmd, timeout=timeout)
+        if debug:
+            logging.debug(output)
+        if record:
+            ret_name = ''.join(('-'.join((self.name, 'test', stamp)), '.log'))
```
Comment (hereischen) [inline provider/storage_benchmark.py]: `'%s-test-%s.log' % (self.name, stamp)` 
is easier to read, and you probably dont need the `random_string`.

**Comment by YongxueHong** [comment]
@CongLi  @XueqiangWei  pls help review thanks.

File: provider/storage_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+TAR_UNPACK = 'tar'
+
+RANDOM_STRING = utils_misc.generate_random_string(8)
+STAMP = ''.join((time.strftime('%Y-%m-%d-%H-%M-%S'), RANDOM_STRING))
+
+
+class StorageBenchmark(object):
+    """
+    Create a Benchmark class which provider common interface(method) for
```
Comment (CongLi) [inline provider/storage_benchmark.py]: s/provider/provides

File: provider/storage_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+        :param timeout: timeout for waiting
+        :type timeout: float
+        """
+        logging.info('Checking the currently running %s processes.' % self.name)
+        if not utils_misc.wait_for(
+                lambda: self.name not in self.session.cmd_output(
+                    self._list_pid % self.name), timeout, step=3.0):
+            raise exceptions.TestError(
+                'All %s processes are not done in %s sec.' % (self.name, timeout))
```
Comment (CongLi) [inline provider/storage_benchmark.py]: Not all processes done in %s sec.

File: provider/storage_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+    def _kill_procs(self):
+        """Kill the specified processors by force."""
```
Comment (CongLi) [inline provider/storage_benchmark.py]: s/processors/processes

File: provider/storage_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+        :param name: the name of benchmark
+        :type name: str
+        """
+        self.session = session
+        self.name = name
+        self.os_type = os_type
+        self.env_files = []
+
+    def __getattr__(self, item):
+        for k, v in self.cmds[self.os_type].items():
```
Comment (hereischen) [inline provider/storage_benchmark.py]:
```python
if item in self.cmds[self.os_type]:
   return self.cmds[self.os_type][item]
```

File: provider/storage_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+
+        :param src: source file
+        :type src: str
+        :param dst: destination directory
+        :type dst: str
+        :param timeout: timeout for installing
+        :type timeout: float
+        """
+        if self.os_type == 'linux':
+            cmd_configure = 'cd %s && ./configure --prefix=%s' % (src, dst)
```
Comment (hereischen) [inline provider/storage_benchmark.py]:
```python
cmd = " cd %s && ./configure --prefix=%s && make && make install" % (src, dst)
```  
The last `join` is kind of overkill.

File: provider/storage_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+
+from avocado.core import exceptions
+
+GIT_DOWNLOAD = 'git'
+CURL_DOWNLOAD = 'curl'
+
+TAR_UNPACK = 'tar'
+
+RANDOM_STRING = utils_misc.generate_random_string(8)
+STAMP = ''.join((time.strftime('%Y-%m-%d-%H-%M-%S'), RANDOM_STRING))
```
Comment (CongLi) [inline provider/storage_benchmark.py]: no use of STAMP, could remove it

File: provider/storage_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+            self.cmd = 'cd %s/src/current && make %s' % (self.iozone_dir, self.arch)
+            self.link_src = '%s/src/current/iozone' % self.iozone_dir
+            self.iozone_path = '/usr/local/bin/iozone'
+            self.setups = {'download_benchmark': (CURL_DOWNLOAD,
+                                                  self.download_url,
+                                                  self.download_path),
+                           'unpack_file': (TAR_UNPACK, self.download_path,
+                                           '/home/iozone_inst'),
+                           'session_cmd': (self.cmd, 300),
+                           'make_symlinks': (self.link_src, self.iozone_path)}
```
Comment (CongLi) [inline provider/storage_benchmark.py]: seems useless to create symlink between source code and dst.

File: provider/storage_benchmark.py [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+class Iozone(StorageBenchmark):
+    def __init__(self, params, session):
+        super(Iozone, self).__init__(params['os_type'], session, 'iozone')
+        self.cfg = IozoneCfg(params['os_type'], params, self)
+        for method, args in self.cfg.setups.items():
+            getattr(self, method)(*args)
+
+    def run(self, cmd_options='-a', timeout=1800):
+        """
+        Run iozone test inside linux guest.
```
Comment (CongLi) [inline provider/storage_benchmark.py]: maybe typo, should support both windows and linux.

File: provider/storage_benchmark.py [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+class Fio(StorageBenchmark):
+    def __init__(self, params, session):
+        super(Fio, self).__init__(params['os_type'], session, 'fio')
+        self.cfg = FioCfg(params['os_type'], params, self)
+        for method, args in self.cfg.setups.items():
+            getattr(self, method)(*args)
+
+    def run(self, cmd_options, timeout=1800):
+        """
+        Run fio test inside linux guest.
```
Comment (CongLi) [inline provider/storage_benchmark.py]: same as iozone, should be typo for both windows and linux guests.

File: provider/storage_benchmark.py [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+        logging.info('Checking the currently running %s processes.' % self.name)
+        if not utils_misc.wait_for(
+                lambda: self.name not in self.session.cmd_output(
+                    self._list_pid % self.name), timeout, step=3.0):
+            raise exceptions.TestError(
+                'All %s processes are not done in %s sec.' % (self.name, timeout))
+
+    def _kill_procs(self):
+        """Kill the specified processors by force."""
+        logging.info('Killing all %s processes by force.' % self.name)
```
Comment (CongLi) [inline provider/storage_benchmark.py]: Should judge if processes existed at first, 
or the cmd execution would be failed if process not existed.
# killall -9 qemu-kvm
qemu-kvm: no process found
# echo $?
1

File: provider/storage_benchmark.py (hunk 13 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 8 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 11 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 9 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 10 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 12 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 14 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 15 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 16 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: thanks, have updated.

File: provider/storage_benchmark.py (hunk 17 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]:
```
>>> session.cmd_output('unamed -sr')
"bash: unamed: command not found...\nSimilar command is: 'uname'\n"
>>> session.cmd('unamed -sr')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/vmt/kar/workspace/lib/python3.6/site-packages/aexpect/client.py", line 1288, in cmd
    raise ShellCmdError(cmd, s, o)
aexpect.exceptions.ShellCmdError: Shell command failed: 'unamed -sr'    (status: 127,    output: "bash: unamed: command not found...\nSimilar command is: 'uname'\n")
>>> 

It's not raise errors.
```

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by xuwei@redhat.com

File: provider/storage_benchmark.py [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+        :param session: session connected guest
+        :param name: the name of benchmark
+        :type name: str
+        """
+        self.session = session
+        self.name = name
+        self.os_type = os_type
+        self.env_files = []
+
+    def __getattr__(self, item):
```
Comment (lolyu) [inline provider/storage_benchmark.py]: for `item` not in `['_symlinks', '_list_pid', '_kill_pid', '_rm_file']`, `__getattr__` should raise `AttributeError`.

File: provider/storage_benchmark.py [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+        :param mode: the mode of unpacking, e.g, tar, unzip
+        :type mode: str
+        :param src: source file
+        :type src: str
+        :param dst: destination directory
+        :type dst: str
+        :param timeout: timeout for unpacking
+        :type timeout: float
+        """
+        getattr(self, self.unpack_cmds[mode])(*(src, dst, timeout))
```
Comment (lolyu) [inline provider/storage_benchmark.py]: the unpack is redundant.

File: provider/storage_benchmark.py [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+        self.setups = {'session_cmd': (self.cmd, 300)}
+
+
+class Iozone(StorageBenchmark):
+    def __init__(self, params, session):
+        self.os_type = params['os_type']
+        super(Iozone, self).__init__(self.os_type, session, 'iozone')
+        self.cfg_map = {'linux': IozoneLinuxCfg, 'windows': IozoneWinCfg}
+        self.cfg = self.cfg_map[self.os_type](params, session)
+        for method, args in self.cfg.setups.items():
```
Comment (lolyu) [inline provider/storage_benchmark.py]: This implicitly assumes dictionary stores items in order `('download_benchmark', 'unpack_file', 'session_cmd')`. Better to stores it in a tuple.

File: provider/storage_benchmark.py (hunk 19 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: Thanks, have updated.

File: provider/storage_benchmark.py (hunk 18 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: Thanks, have updated.

Comment (YongxueHong) [reply provider/storage_benchmark.py]: Thanks, have updated.

File: provider/storage_benchmark.py (hunk 20 above)
Comment (YongxueHong) [reply provider/storage_benchmark.py]: Thanks, have updated.

File: provider/storage_benchmark.py [hunk 21]
```diff
@@ -0,0 +1,335 @@
...
+import logging
+import os
+import re
+from collections import OrderedDict
+
+from platform import machine
+
+from virttest import utils_misc
+
+from avocado.core import exceptions
```
Comment (lolyu) [inline provider/storage_benchmark.py]: `from avocado import TestError`

Comment (YongxueHong) [reply provider/storage_benchmark.py]: Thanks a lot. have updated.

**Comment by CongLi** [comment]
Ack.

File: provider/storage_benchmark.py [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+        self.cfg = self.cfg_map[self.os_type](params, session)
+        for method, args in self.cfg.setups.items():
+            getattr(self, method)(*args)
+
+    def run(self, cmd_options, timeout=1800):
+        """
+        Run fio test inside guest.
+
+        :param cmd_options: fio command options, e.g, --filename=/home/test
+                            --direct=1 --rw=read --bs=64K --size=1000M
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply provider/storage_benchmark.py]: ok, I will state it clearly.

**Comment by YongxueHong** [comment]
Thanks all.

**Pull request merged**